### DISCUSSION TITLE
kf-tools--pipelines ended in degraded in my first local quickstart

### DISCUSSION THREAD

**Discussion by cheese** [deployKF/deployKF#113] [Questions & Answers] [answered]
I am truly a newbie to Kubeflow and deployKF. I want to deploy my first Kubeflow with deployKF to get started. I followed https://www.deploykf.org/guides/local-quickstart to set up an instance of deployKF [v0.1.4 (29ac97a)](https://github.com/deployKF/deployKF/commit/29ac97a8727d4da10f456263d755e391faface2d) on my Fedora 38 machine with moby-engine 24.0.5 and k3d 5.6.0, but it ended up with `kf-tools--pipelines` in degraded state.

Four pods are failing to run:
1. `ml-pipeline` and `ml-pipeline-ui` pods error: secret "cloned--kubeflow-pipelines--backend-object-store-auth" not found
2. `ml-pipeline-ui-artifact` pods error: secret "mlpipeline-minio-artifact" not found

**Comment by thesuperzapper** [comment]
@cheese can you check the state AND logs of the Pods in the `kyverno` namespace, especially the `kyverno-background-controller-xxxxxxx` Pod? 

After you save the `background-controller` logs (so you can post them), just delete the Pod:

```bash
kubectl delete pods -n kyverno -l "app.kubernetes.io/part-of=kyverno,app.kubernetes.io/component=background-controller"
```

After it comes back up, check that the following secrets were created:

- `Secret/cloned--kubeflow-pipelines--backend-object-store-auth` (in `kubeflow` namespace)
- `Secret/mlpipeline-minio-artifact` (in `team-1` namespace)

**Comment by cheese** [comment]
Logs of `background-controller` are scrolling fast, I just cut the tail part. 
[kyverno-kyverno-background-controller-796f77c79f-hs267-1710737551788174593.log](https://github.com/deployKF/deployKF/files/14631152/kyverno-kyverno-background-controller-796f77c79f-hs267-1710737551788174593.log)

I deleted the pod but the secrets were not created after it came back.

**Reply by thesuperzapper** [reply, marked as answer]
@cheese it really looks like something deleted the minio secrets which are the source of these clones:

- `Secret/generated--kubeflow-pipelines--backend-object-store-auth` (in the `deploykf-minio` namespace)
- `Secret/generated--kubeflow-pipelines--profile-object-store-auth--team-1` (in the `deploykf-minio` namespace)
- `Secret/generated--kubeflow-pipelines--profile-object-store-auth--team-1-prod`  (in the `deploykf-minio` namespace)

Did you delete these manaully? 
___(NOTE:__ they show as "need pruning" in the ArgoCD UI, which might be confusing because they should NOT be deleted, they aren't created by ArgoCD, but instead by a `pre-install` Job Pod)_


Either way, you can trigger the job to recreate them by re-syncing the `dkf-opt--deploykf-minio` ArgoCD app manually, even though it will show as "in sync", you can __do this in the ArgoCD WebUI__. 

If you need help accessing the ArgoCD WebUI, see the [sync apps](https://www.deploykf.org/guides/getting-started/#sync-argocd-applications) step of the getting started guide.

**Reply by cheese** [reply]
Thanks! I remember I did not delete them. After that, re-syncing `dkf-opt--deploykf-minio` just fixed the issue.

**Reply by thesuperzapper** [reply]
@cheese if possible, could you try and reproduce this situation in a new K3D cluster?

Because there is clearly some race condition (or bug) here with the `dkf-opt--deploykf-minio` app, and I want to figure out what it is.

**Reply by cheese** [reply]
I tried again and still had to re-sync `dkf-opt--deploykf-minio` after it announced healthy.

**Reply by thesuperzapper** [reply]
@cheese it is interesting that it's reproducible.

Can you tell me the version of the `argocd` CLI you are running? Run `argocd version` to check.

After that, run `brew update` and `brew upgrade` to update to the latest one.

Finally, can you please try with a new cluster and the latest version of ArgoCD and see if it still happens? 
(I have updated the version of ArgoCD to `2.10.4` in `main` branch of `deploykf/deploykf`, so you should just be able to re-clone the repo following the [prepare argocd](https://www.deploykf.org/guides/local-quickstart/#3-prepare-argocd) step in the local quickstart again).

**Reply by cheese** [reply]
My argocd version was 2.10.3.
```
$ argocd version                                                                                                                                                       
argocd: v2.10.3+0fd6344                                                                                                                                                                              
  BuildDate: 2024-03-13T19:37:04Z                                                                                                                                                                    
  GitCommit: 0fd6344537eb948cff602824a1d060421ceff40e                                                                                                                                                
  GitTreeState: clean                                                                                                                                                                                
  GoVersion: go1.21.7                                                                                                                                                                                
  Compiler: gc                                                                                                                                                                                       
  Platform: linux/amd64
```

I tried again with argocd CLI 2.10.4 and your updated script, but `istiod` pod failed to run. Tail log of `istiod` 
[istio-system-istiod-7b9b6df595-8s94b-1710814760490227083.log](https://github.com/deployKF/deployKF/files/14644483/istio-system-istiod-7b9b6df595-8s94b-1710814760490227083.log)

**Reply by thesuperzapper** [reply]
@cheese just so we are clear:

1. did you delete and recreate the k3d cluster (or just delete some of the things before re-syncing)? 
2. did you use the new `./deploykf/argocd-plugin/install_argocd.sh` to upgrade your ArgoCD on the cluster?

Because those errors look like Kyverno is not running properly at the time.
If you dont recreate the cluster, you have to be careful removing deployKF, see the [uninstall guide](https://www.deploykf.org/guides/uninstall/).

But either way, can we try again with a clean cluster, and see if the MinIO error happens again?

**Reply by cheese** [reply]
Yes to both. I am too new to KubeFlow, so I always delete and recreate the k3d cluster.

I tried again with argocd 2.9.6 but it still failed at running the `istiod` pod.

**Reply by thesuperzapper** [reply]
@cheese that makes no sense, are you actually seeing the same error than you raised in https://github.com/orgs/deployKF/discussions/113#discussioncomment-8834965 on the `istiod` pod?

If you are seeing anything like `denied the request: mutation policy kubeflow-pipelines--clone-mysql-secret error: failed to validate resource mutated by policy kubeflow-pipelines--clone-mysql-secret: unknown object type "nil" in io.k8s.api.core.v1.Secret.data.cert-chain.pem` in your `istiod` logs, it means you are not installing into a clean cluster, the Kyverno resources are synced AFTER the Istio ones, (if you use the sync script), so they can't be there.

Use `kubectx` to verify that you have the `kubectl` context that you think selected:

```bash
# `fzf` makes `kubectx` interative
brew install kubectx fzf

# then run this command, and use the interactive selector to choose `k3d-deploykf`
kubectx
```

Just to confirm, you are running a command like this to create a k3d cluster?

```bash
# delete the existing cluster
k3d cluster delete "deploykf"

# create a new clean cluster
k3d cluster create "deploykf" \
  --image "rancher/k3s:v1.27.10-k3s2"
```

Once you do that, you can verify that only the cluster Pods are running:

```bash
> kubectl get pods --all-namespaces

NAMESPACE     NAME                                     READY   STATUS      RESTARTS   AGE
kube-system   coredns-77ccd57875-zqdw6                 1/1     Running     0          12s
kube-system   local-path-provisioner-957fdf8bc-tfr2w   1/1     Running     0          12s
kube-system   metrics-server-648b5df564-kwjc4          0/1     Running     0          12s
kube-system   helm-install-traefik-crd-599wh           0/1     Completed   0          12s
kube-system   svclb-traefik-5068d407-6vb9r             2/2     Running     0          4s
kube-system   traefik-768bdcdcdd-bw4rn                 0/1     Running     0          4s
kube-system   helm-install-traefik-xvclv               0/1     Completed   1          12s
```

Then make sure you are using the latest `main` versions of `./deploykf/argocd-plugin/install_argocd.sh` and `./deploykf/scripts/sync_argocd_apps.sh` and follow [local quickstart](https://www.deploykf.org/guides/local-quickstart/#3-prepare-argocd) again:

```bash
# update your argod cli
brew upgrade argocd

# remove the deploykf/ directory before re-cloning
rm -rf deploykf

# re clone
git clone -b main https://github.com/deployKF/deployKF.git ./deploykf

# install argocd
# NOTE: wait for it to be ready
bash ./deploykf/argocd-plugin/install_argocd.sh

# apply the app-of-apps
# NOTE: you can also do this manually, but this command applies the default 0.1.4 app-of-apps with sample-values.yaml
#        https://github.com/deployKF/deployKF/blob/main/argocd-plugin/example-app-of-apps/app-of-apps.yaml
bash ./deploykf/argocd-plugin/example-app-of-apps/apply_apps_of_apps.sh

# sync apps
bash ./deploykf/scripts/sync_argocd_apps.sh
```

**Reply by cheese** [reply]
I tried again with ArgoCD 2.10.4 and it turned all green. Minio issue is gone. This time I only used the sync script to sync argocd apps. I synced argocd with Web GUI and failed in `istiod`, since I didn't know Istio has to be synced before Kyverno.

**Reply by thesuperzapper** [reply]
@cheese out of interest were you trying to sync them all at once, or were you at least doing it by group?

**Reply by cheese** [reply]
I am also new to argocd. I synced all apps at once when I was using the Web GUI.

**Reply by thesuperzapper** [reply]
@cheese thanks for the feedback, I have put a warning on the [Local Quickstart](https://www.deploykf.org/guides/local-quickstart/#5-sync-argocd-applications) saying not to sync everything at once, to help future users.
